### PR TITLE
ADD: First draft of comparrison

### PR DESCRIPTION
**Pull request by madisonmay**

### CODE REVIEW

**Comment by madisonmay** [comment]
Not yet ready for review

File: finetune/base.py [hunk 1]
```diff
@@ -149,32 +149,35 @@ def _eval(self, *tensors, feed_dict):
...
+                                                              random_state=self.config.seed)
+        train_xs = list(zip(*train_xs))
+        test_xs = list(zip(*test_xs))
 
-        self._build_model(n_updates_total=n_updates_total, target_dim=target_dim)
+        train_x, train_mask = self._text_to_ids(*train_xs)
+        test_x, test_mask = self._text_to_ids(*test_xs)
 
-        dataset = (train_x, train_mask, Y)
+        return self._training_loop(train_x, train_mask, train_y, test_x, test_mask, test_y, batch_size)
```
Comment (madisonmay) [inline finetune/base.py]: pass as kwargs

File: finetune/comparison.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        forward_pairs = self.encoder.encode_multi_input(X1, X2, max_length=max_length, verbose=self.config.verbose)
+        backward_pairs = self.encoder.encode_multi_input(X2, X1, max_length=max_length, verbose=self.config.verbose)
+        seq_array_fw = self._array_format(forward_pairs)
+        seq_array_bw = self._array_format(backward_pairs)
+        token_ids = np.stack([seq_array_fw.token_ids, seq_array_bw.token_ids], 1)
+        mask = np.stack([seq_array_fw.mask, seq_array_bw.mask], 1)
+
+        return token_ids, mask
+
+    def finetune(self, X1, X2, Y, batch_size=None):
```
Comment (madisonmay) [inline finetune/comparison.py]: X_1 vs X1

File: finetune/comparison.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+    def _define_placeholders(self):
+        super()._define_placeholders()
+        self.X = tf.placeholder(tf.int32, [None, 2, self.config.max_length, 2])
+        self.M = tf.placeholder(tf.float32, [None, 2, self.config.max_length])  # sequence mask
+
+    def _target_model(self, featurizer_state, targets, n_outputs, train=False, reuse=None, **kwargs):
+        featurizer_state["sequence_features"] = tf.reduce_sum(featurizer_state["sequence_features"], 1)
+        featurizer_state["features"] = tf.reduce_sum(featurizer_state["features"], 1)
+        return super()._target_model(featurizer_state, targets, n_outputs, train, reuse, **kwargs)
```
Comment (madisonmay) [inline finetune/comparison.py]: pass as kwargs

File: finetune/datasets/quora_similarity.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+            text_column1="question1",
+            text_column2="question2",
+            target_column="is_duplicate",
+            filename=QUORA_SIMILARITY
+        )
+
+if __name__ == "__main__":
+    # Train and evaluate on SST
+    dataset = QuoraDuplicate(nrows=50000).dataframe
+    print(dataset)
```
Comment (benleetownsend) [inline finetune/datasets/quora_similarity.py]: oops.

File: finetune/comparison.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        return token_ids, mask
+
+    def finetune(self, X_1, X_2, Y, batch_size=None):
+        """
+        :param Xs: An iterable of lists or array of text, shape [batch, n_inputs, tokens]
+        :param Y: integer or string-valued class labels. It is necessary for the items of Y to be sortable.
+        :param batch_size: integer number of examples per batch. When N_GPUS > 1, this number
+                           corresponds to the number of training examples provided to each GPU.
+        """
+        if self.target_type is None:
```
Comment (benleetownsend) [inline finetune/comparison.py]: I didnt know whether it was desired to support regression for this?

File: finetune/datasets/__init__.py [hunk 6]
```diff
@@ -11,7 +11,8 @@ def file_hash(path_obj):
     with path_obj.open(mode='rb', buffering=0) as f:
         for b in iter(lambda: f.read(128 * 1024), b''):
             h.update(b)
-    return h.hexdigest()
+    print("MD5: {}".format(h.hexdigest()))
```
Comment (benleetownsend) [inline finetune/datasets/__init__.py]: Can switch for a logger if you think its worth it.

File: finetune/datasets/__init__.py [hunk 7]
```diff
@@ -60,3 +60,22 @@ def generic_download(url, text_column, target_column, filename, save=True, task_
         new_df.to_csv(save_path, index=False)
 
     return new_df
+
+
+def comparison_download(url, text_column1, text_column2, target_column, filename, save=True, task_type='Similarity', data_directory='Data'):
```
Comment (benleetownsend) [inline finetune/datasets/__init__.py]: Can combine with the other download_fn if you can tell me what interface youd like.

File: tests/test_comparison.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+    def test_reasonable_predictions(self):
+        model = Comparison(config=self.default_config(), n_epochs=3)
+        animals = ["The cat says Meow.",
+                   "The chicken says sqwark.",
+                   "The fish says bloop.",
+                   "The dog says Woof.",
+                   "The frog says \"ribbet\"."]
+
+        prince_philip_quotes = ["When a man opens a car door for his wife, it's either a new car or a new wife.",
```
Comment (benleetownsend) [inline tests/test_comparison.py]: Im going to switch these out for something simpler that will hopefully yield better results. Apparently this task is pretty hard.

File: finetune/datasets/__init__.py (hunk 7 above)
Comment (madisonmay) [reply finetune/datasets/__init__.py]: could standardize on input_columns

File: finetune/network_modules.py [hunk 9]
```diff
@@ -59,8 +65,8 @@ def featurizer(X, encoder, dropout_placeholder, config, train=False, reuse=None,
         clf_h = tf.reshape(clf_h, [-1, config.n_embed])  # [batch, embed]
         return {
             'embed_weights': embed_weights,
-            'features': clf_h,
-            'sequence_features': h
+            'features': tf.reshape(clf_h, shape=initial_shape[0: -2] + [config.n_embed]),
```
Comment (madisonmay) [inline finetune/network_modules.py]: Just move logic up a bit out of return statement

File: finetune/comparison.py (hunk 5 above)
Comment (benleetownsend) [reply finetune/comparison.py]: Ill just assume this is all tided up in your PR #64  and will leave this for now.

**Pull request merged**